certora/ArrangerConduit.spec

```diff
@@ -0,0 +1,805 @@
+    mathint numRequestsAfter = getFundRequestsLength();
+
+    mathint requestedFundsOtherAfter = requestedFunds(otherAsset, otherIlk);
+    mathint totalRequestedFundsOtherAfter = totalRequestedFunds(otherAsset2); 
+
+    IArrangerConduit.FundRequest requestAfter = getFundRequest(anyIndex);
+    bytes32 infoHashAfter = aux.hashString(requestAfter.info);
+
+    assert numRequestsAfter == numRequestsBefore, "num requests changed";
+    assert anyIndex == fundRequestId => requestedFundsAfter == requestedFundsBefore - requestBefore.amountRequested, "cancelFundRequest did not decrease by amount";
+    assert anyIndex == fundRequestId => totalRequestedFundsAfter == totalRequestedFundsBefore - requestBefore.amountRequested, "totalRequestedFunds did not decrease by amount";
+    assert anyIndex == fundRequestId => requestAfter.status == IArrangerConduit.StatusEnum.CANCELLED, "cancelFundRequest did not change status to CANCELLED";
+    assert anyIndex != fundRequestId => requestAfter.status == requestBefore.status, "cancelFundRequest on another index changed status";
+    assert requestAfter.asset == requestBefore.asset
+        && requestAfter.ilk == requestBefore.ilk
+        && requestAfter.amountRequested == requestBefore.amountRequested
+        && requestAfter.amountFilled == requestBefore.amountFilled
+        && infoHashAfter == infoHashBefore,
+        "other request params not as before";
+
+    assert anyIndex == fundRequestId => requestedFundsOtherAfter == requestedFundsOtherBefore, "other requested funds changed unexpectedly";
+    assert anyIndex == fundRequestId => totalRequestedFundsOtherAfter == totalRequestedFundsOtherBefore, "other total requested funds changed unexpectedly";
+}
+
+// TODO: figure out why this is still not working
+// Verify revert rules on cancelFundRequest
+rule cancelFundRequest_revert(uint256 fundRequestId) {
+    env e;
+
+    IArrangerConduit.FundRequest request = getFundRequest(fundRequestId);
+
+    require request.asset == gem;
+
+    bool canCall = roles.canCall(request.ilk, e.msg.sender, currentContract, to_bytes4(0x933d9476)); // cancelFundRequest(uint256)
+    mathint requestedFunds = requestedFunds(request.asset, request.ilk);
+    mathint totalRequestedFunds = totalRequestedFunds(request.asset);
+    
+    cancelFundRequest@withrevert(e, fundRequestId);
+
+    bool revert1 = e.msg.value > 0;
+    bool revert2 = !canCall;
+    bool revert3 = request.status != IArrangerConduit.StatusEnum.PENDING;
+    bool revert4 = requestedFunds < to_mathint(request.amountRequested);
+    bool revert5 = totalRequestedFunds < to_mathint(request.amountRequested);
+
+    assert lastReverted <=> revert1 || revert2 || revert3 ||
+                            revert4 || revert5, "Revert rules failed";
+
+}
+
+// Verify correct storage changes for non reverting requestFunds
+rule drawFunds(address asset, address destination, uint256 amount) {
+    env e;
+
+    require asset == gem;
+    require currentContract != destination;
```

Comment:
This is a bit of personal taste, but I prefer to put requires when we are sure some address can't be other. But in this case is not even a manageable address, is one being passed by parameter, so I'd probably avoid this require and I'd check the assert for both cases.

Reply:
Done